pkg/storage/chunk/aws/s3_storage_client.go

```diff
@@ -392,22 +392,34 @@ func (a *S3ObjectClient) GetObject(ctx context.Context, objectKey string) (io.Re
 
 // PutObject into the store
 func (a *S3ObjectClient) PutObject(ctx context.Context, objectKey string, object io.ReadSeeker) error {
-	return instrument.CollectedRequest(ctx, "S3.PutObject", s3RequestDuration, instrument.ErrorCode, func(ctx context.Context) error {
-		putObjectInput := &s3.PutObjectInput{
-			Body:   object,
-			Bucket: aws.String(a.bucketFromKey(objectKey)),
-			Key:    aws.String(objectKey),
+	retries := backoff.New(ctx, a.cfg.BackoffConfig)
+	err := ctx.Err()
+	for retries.Ongoing() {
+		if ctx.Err() != nil {
+			return errors.Wrap(ctx.Err(), "ctx related error during s3 putObject")
 		}
+		err = instrument.CollectedRequest(ctx, "S3.PutObject", s3RequestDuration, instrument.ErrorCode, func(ctx context.Context) error {
+			putObjectInput := &s3.PutObjectInput{
+				Body:   object,
+				Bucket: aws.String(a.bucketFromKey(objectKey)),
```

Comment:
you can save the result of `a.bucketFromKey(objectKey)` to a variable outside the retry loop so we don't make this call every time